### PR TITLE
Fix and standardize multiple table UI inconsistencies

### LINKED ISSUE #4500: Feature flags table formatted wrong on the first load
## Bug description

The first time I load the feature flags page (with a long description), it displays weirdly. After refreshing, it goes away.

1st Load:
image

Subsequent Load:
image

*Please describe.*  
*If this affects the front-end, screenshots would be of great help.*  

## Expected behavior

It loads correctly formatted all the time.

## How to reproduce

1. Create a feature flag with a long description
2. Load the feature flag page

## Environment

- [ ] PostHog Cloud
- [x] self-hosted PostHog, version/commit: 1.25.0

## Additional context

None.

### LINKED ISSUE #4721: Squishy Tables
## Bug description

Tables are not too readable on small mobile devices

Screen Shot 2021-06-11 at 3 16 20 PM

## Expected behavior

Scrolly table

https://user-images.githubusercontent.com/13460330/121753944-3e27d580-cac8-11eb-9f38-902f6660cedc.mov

## How to reproduce

1. Go to a Feature Flags or Annotations page
2. Make window a lot smaller

## Environment

- [X] PostHog Cloud
- [X] self-hosted PostHog, version/commit: _please provide_

### LINKED ISSUE #4723: Actions Table Overflow inaccessible on smaller screens
## Bug description

Columns to the right are hidden and I can't scroll to access them.

Screen Shot 2021-06-11 at 5 05 25 PM

## Expected behavior

Allow scrolling.

## How to reproduce

1. Go to actions table
2. Make window smaller
3. Claustrophobia

## Environment

- [ ] PostHog Cloud
- [ ] self-hosted PostHog, version/commit: _please provide_

### LINKED ISSUE #4724: Inconsistent scroll shading for fully visible table
## Bug description

Even when the event table is fully in view, we see weird shading business in our events table. This doesn't happen anywhere else in our app (afaik).

Smaller screen
Screen Shot 2021-06-11 at 5 12 56 PM

## Expected behavior

Make table overflow behavior consistent across pages.

## How to reproduce

1. Go to events table
2. Play around with window sizes.

## Environment

- [X] PostHog Cloud
- [X] self-hosted PostHog, version/commit: _please provide_

### LINKED ISSUE #4725: Table Overflow Revamp
## Bug description

It's getting hard to keep track of all the open issues about table bugginess across our app, so this issue is the master ticket!

Related bugs

- #4500 Feature flags table formatted wrong on the first load
- #4721 Squishy tables
- #4723 Actions Table Overflow inaccessible on smaller screens
- #4724 Inconsistent scroll shading for fully visible table

## Changes expected
- Fixed table width that allows scrolling to access entire table on mobile
  - standardize table overflow handling across ALL of our tables for a consistent UX
  - reduces paint frequency -> less laggy
- Text overflow handled with ellipses

## Environment

- [X] PostHog Cloud
- [X] self-hosted PostHog, version/commit: _please provide_

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Wrapped up a bunch of table fixes in this PR.

- Fixes #4725 (Master) Table Overflow Revamp
- Fixes #4500 Feature flags table formatted wrong on the first load
- Fixes #4721 Squishy tables
- Fixes #4723 Actions Table Overflow inaccessible on smaller screens
- Fixes #4724 Inconsistent scroll shading for fully visible table

## A bunch of after pics (in no particular order)

Screen Shot 2021-06-11 at 4 39 55 PM

https://user-images.githubusercontent.com/13460330/121764458-efdefa80-caf8-11eb-90c5-c7304560a192.mov

https://user-images.githubusercontent.com/13460330/121764445-c6be6a00-caf8-11eb-87c9-130563d65905.mov

Screen Shot 2021-06-11 at 8 55 22 PM

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [X] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] Frontend/CSS is usable at 320px (iPhone SE) and decent at 360px (most phones)

### CODE REVIEW

File: frontend/src/lib/utils/responsiveUtils.ts [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
-import { ANTD_EXPAND_BUTTON_WIDTH } from './index'
+import { ANTD_EXPAND_BUTTON_WIDTH } from '../components/ResizableTable'
+
+const BREAKPOINT_MAP = Object.entries(responsiveMap).reduce<Record<string, number>>(
+    (acc, [key, cssStatement]) => ({
+        ...acc,
+        [key]: parsePixelValue(cssStatement),
+    }),
+    {}
+)
```
Comment (mariusandra) [inline frontend/src/lib/utils/responsiveUtils.ts]: Let's avoid using `reduce` whenever possible. The only thing using `reduce` shows to whoever is reading the code is that the author is proficient in using `reduce`, _nothing_ else :). It's usually longer and always less readable (`==` maintainable) than all other ways of doing the same thing.

The following does the same, but is much simpler to understand:

```js
const BREAKPOINT_MAP = Object.fromEntries(
    Object.entries(responsiveMap).map(([key, cssStatement]) => [key, parsePixelValue(cssStatement)])
)
```

... and is most likely much faster as well (no need to create and spread `array.length` temporary objects).

File: frontend/src/lib/utils/responsiveUtils.ts [hunk 2]
```diff
@@ -16,12 +25,15 @@ export function parsePixelValue(cssStatement: string): number {
...
-    let breakpoint = breakpoints[0]
-    breakpoints.forEach((value) => {
+    let breakpoint = BREAKPOINT_VALUES[0]
+    BREAKPOINT_VALUES.forEach((value) => {
         if (width > breakpoint) {
             breakpoint = value
         }
     })
     return breakpoint
 }
```
Comment (mariusandra) [inline frontend/src/lib/utils/responsiveUtils.ts]: I'd also simplify this to:

```js
export function getActiveBreakpoint(): number {
    const width = window.innerWidth
    const lastMatchingBreakpoint = BREAKPOINT_VALUES.filter(value => width >= value).pop()
    return lastMatchingBreakpoint || BREAKPOINT_VALUES[0]
}
```

Comment (mariusandra) [reply frontend/src/lib/utils/responsiveUtils.ts]: I also renamed this function `getActiveBreakpointValue` (over `getActiveBreakpoint`) since technically a "breakpoint" is one of "sm", "xs", etc, not the value in pixels it resolves to.

And I changed `width > value` to `width >= value`, as that's how (antd) breakpoints [are defined](https://ant.design/components/grid/#Col).

**Review by mariusandra — CHANGES_REQUESTED** [review]
I simplified the breakpoint code a bit, yet still found several issues in the app that I'll gladly pass on :)

1. Feature flags do not display well, no matter how I open the page. The key is always cut for me, and the description area is needlessly huge:
image

2. Events page (zoomed out to simulate a wide screen) won't let me resize things like I expect:


3. No "you can scroll right" indicator here:


Probably outside the scope, or maybe not:

4. If the widths auto-update while resizing, my dragging cursor is "let go" and I need to start resizing again:



5. There's no way to make the last column of any table wider or shorter, plus the whitespace that's at the end of the table:


---

PS, my most common PR review comment is "simplify!", so feel free to always do that automatically :D. The logic here is that code is *not* written for computers to read, but for other humans to read. Computers can read your code no matter how you well you write it, but the human reading it will be caught off guard by clever code that's hard to parse. So always optimise code clarity for the cognitively lowest life form in the process: your fellow humans. Usually this means "make your code so simple it looks like a first year student wrote it".

**Comment by mariusandra** [comment]
The breaking cypress tests seem relevant as well.

**Comment by mariusandra** [comment]
One more bug, this time in [this](https://app.posthog.com/insights?insight=TRENDS&interval=day&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%7D%5D&filter_test_accounts=false&properties=%5B%5D&display=ActionsTable) cohorts table:

**Comment by paolodamico** [comment]
Looks like @mariusandra already did a comprehensive review of this, tag me for a follow up review if needed @alexkim205

**Comment by samwinslow** [comment]
> 2. Events page (zoomed out to simulate a wide screen) won't let me resize things like I expect:

Fixed this by removing the arbitrary max column width I had set by default.

> 1. There's no way to make the last column of any table wider or shorter, plus the whitespace that's at the end of the table:

This one gets tricky because in order for the table to render correctly, one column needs to have an uncontrolled width set by the browser. But, because the size of the entire table and other columns update together, it works out such that this "uncontrolled" column appears to have a fixed width. Right now the last column is chosen for this unfortunate caveat, but maybe we could always push a blank column onto the end of the ones given to us via props.

**Comment by mariusandra** [comment]
> In order for the table to render correctly, one column needs to have an uncontrolled width set by the browser.

Why? :)

**Comment by samwinslow** [comment]
@mariusandra If every column has a set width, the browser will just decide what it thinks appropriate width values are in order to fill the horizontal space of the `<table>`, when using default `table-layout`... Antd's `tableLayout: fixed` prop did not seem to change this behavior but more experimentation could be done. 

Antd's docs have a similar warning:
> Specify width of columns if header and cell do not align properly. If specified width is not working or have gutter between columns, please try to leave one column at least without width to fit fluid layout, or make sure no long word to break table layout.

When I was working on ResizableTable it was very hard to keep the resizable column headers in sync with the table's actual columns. I don't mean to suggest it's impossible but it might get hacky!

**Comment by mariusandra** [comment]
Yeah, this sounds tricky. I don't know the nitty gritty details of antd's tables, and those tradeoffs probably make sense.

Yet it's funny how "resizable table headers" is still an unsolved computer science problem :).

Can't we just, after the table has fully loaded, capture all the flexibly set column widths, and then just keep them all in a controlled state as we start to resize? Not sure if this helps anything 😅

File: frontend/src/lib/utils/responsiveUtils.ts (hunk 1 above)
Comment (alexkim205) [reply frontend/src/lib/utils/responsiveUtils.ts]: super clean, thanks for the tip :)

**Comment by alexkim205** [comment]
Thanks for the input everyone. Given the above discussion on ResizableTable, there're two more things I want to look into in this PR. 

1. Bug fixes (thanks @mariusandra for the comprehensive dive)
2. Refactoring + Simplifying `<ResizableTable/>` (which fixes some bugs in 1.)

Please correct me if I'm wrong, but from my interpretation of [today's code](https://github.com/PostHog/posthog/blob/6d89514ad6f06b7d6deb2ecc058dea8b875c189d/frontend/src/lib/components/ResizableTable/index.tsx#L223-L231), we nix antd's table header, and prepend a custom header with columns that we keep in sync with the table's columns. If the above reading of the code is correct, this seems to bring a lot more pain than joy when dealing with edge cases like:

- filling out the last column width
- table scrolling independently from headers
- inconsistent scroll gradient
- more scroll handlers
- and more!

Have we previously considered the resizable columns implementation in the [official antd](https://ant.design/components/table/#components-table-demo-resizable-column) documentation? If I'm not overlooking an obvious disadvantage with this option, I feel passing in our `header.cell` component instead of completely replacing the component and handling column sizing in state ourselves is a lot cleaner and easier to maintain.

---

More on point 2: I'll put out a proof of concept about what I mean soon

**Comment by samwinslow** [comment]
@alexkim205 The example in the Antd docs is a bit of a red herring. It's not an obvious disadvantage at first — it re-renders the entire table on each resize event, which happens every few milliseconds. It gets borked on very long tables like Events or Sessions. Even if the resizes are throttled to occur once every 60 or 100 ms, e.g. with `requestAnimationFrame`, rendering may still take longer, and a 100ms interval works out to about 10 frames per second — looks quite bad.

This actually was my first approach, but then I settled on a CSS-based solution; updating the widths specified in `<colgroup>` styles does not trigger a React render, but only a browser repaint.

**Comment by alexkim205** [comment]
Discussed some possible approaches during standup. Here are some notes for posterity:

- passing in `columns` prop with width makes antd's `<Table>` component rerender aggressively
- Virtual table header is a good solution but introduces several bugs like the one this PR tries to address
- Maybe see if we can use other table libraries that can better handle of large amounts of data
   - [ ] `react-table` - [CodeSandbox](https://codesandbox.io/s/nice-christian-4gkhr) doesn't handle many rows well either.
   - [ ] `react-virtualized` - [CodeSandbox](https://codesandbox.io/s/react-virtualized-resizable-columns-forked-8yie9) is promising.
   - [ ] [`react-window`](https://github.com/bvaughn/react-window) - smaller lightweight version of `react-virtualized`
   - [x] [`react-table`](https://react-table.tanstack.com/docs/examples/sub-components) + `react-window` - highly extensible way to use hooks and custom ui components to bring expandable, virtualization, resizable features to existing table

**Comment by mariusandra** [comment]
`react-table` looks pretty awesome! Unless I'm terribly wrong, it shouldn't be that hard to implement it and make our tables look like they do now? :)

**Comment by alexkim205** [comment]
Yup it's a headless utility library so it's pretty cool that you can pick and choose features you want to include. One feature I've been having a hard time playing nicely is expandability. Still trying to figure out a way to make rows expandable without breaking resizability and virtualization

In summary, want to combine 
* https://react-table.tanstack.com/docs/examples/virtualized-rows
* https://react-table.tanstack.com/docs/examples/expanding
* https://react-table.tanstack.com/docs/examples/full-width-resizable-table <--- example is broken on their website

**Comment by alexkim205** [comment]
> 1. Feature flags do not display well, no matter how I open the page. The key is always cut for me, and the description area is...

1. Addressed bug 1 by ellipsifying key and description columns (the two most variable columns) for more consistent row heights. Added tooltips so that user can still view any text that is cutoff. 
2. Fixed failing cypress tests with more specific selectors
3. Fixed column widths of action columns (Enabled, Usage, Actions) so that we can have more space for the other columns and fixed them to the right.

### Demo


https://user-images.githubusercontent.com/13460330/122113089-7e43cc80-cdd6-11eb-80cf-f7ff3249c327.mov



Bugs 2-5 happen on the events table and I think it requires a different implementation of the events table to fix this. I'll address this in another PR as it's out of the scope.

**Comment by alexkim205** [comment]
> One more bug, this time in [this](https://app.posthog.com/insights?insight=TRENDS&interval=day&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%7D%5D&filter_test_accounts=false&properties=%5B%5D&display=ActionsTable) cohorts table:
> 

I think this bug is fixed by my commit [here](https://github.com/PostHog/posthog/pull/4700/commits/3270d43a66c25f9473ea12f301c0a77254e71159) that was merged in as part of @paolodamico's InsightsTableV2 update. But the fix hasn't been made in V1, so I'll bring my commit over here too.

**Comment by alexkim205** [comment]
Whelp, should have merged instead of rebase from master. Please ignore all those commits that I brought in except the last one. I'll squash and clean up accordingly when I merge this PR in.

**Comment by mariusandra** [comment]
Uh... sorry, but I can't review this. Not sure how it came to be, but there are thousands of changed lines in 77 files:

image

Can you please clean it up? :)

**Review by mariusandra — CHANGES_REQUESTED** [review]
.

**Comment by alexkim205** [comment]
Hey @mariusandra sorry about the commit cereal. I think this rev should be cleaner

File: frontend/src/lib/components/Table.tsx [hunk 3]
```diff
@@ -7,6 +7,7 @@ import { TZLabel } from './TimezoneAware'
 export function createdAtColumn(): Record<string, any> {
     return {
         title: 'Created',
+        ellipsis: true,
```
Comment (mariusandra) [inline frontend/src/lib/components/Table.tsx:10]: Instead of an ellipsis, can't we just "nowrap" this column? The "2 days ago" texts are usually very small, and just having a random "..." in the end feels weird:

image

File: frontend/src/lib/components/Table.tsx [hunk 4]
```diff
@@ -19,6 +20,7 @@ export function createdByColumn(items: Record<string, any>[]): Record<string, an
     const { user } = useValues(userLogic)
     return {
         title: 'Created by',
+        width: 150,
```
Comment (mariusandra) [inline frontend/src/lib/components/Table.tsx:23]: I'm also slightly vary about imposing this. You might have a long (e.g. indian, russian, whatever) name that just breaks with "..." on the last letter. 

I'd be a much bigger fan of something like: "max-width: 300px, width: auto", though that might be annoyingly complicated to implement?

**Review by mariusandra — DISMISSED** [review]
I'm not sure how annoying this feedback is to fix, but it's all given in the service of "let's have better UX". I'm also not sure if any of my comments override anything said before or not. :)

Some comments inline, some below.

1. Let's remove the ellipsis from the feature flags table. I think There's no reason this table needs fixed heights. Looking at [our own data](https://app.posthog.com/feature_flags), if we'd merge this, we'd have a sub-par experience... as we'll have mountains of whitespace all around the page, yet we won't be able to actually read any flag keys and descriptions. That sounds like a failure. So can we just make it look like this, but... uhm... work? :D

image

2. Feature flags don't tick the "320px" checkbox: 


3. Similar comment for the annotations table. The annotation column is anyway in two lines sometimes if "last updated" pushes it so.... so why the ellipsis? Let's remove it. The whitespace is also very suboptimal in full screen.

image
image

File: frontend/src/lib/components/Table.tsx (hunk 3 above)
Comment (alexkim205) [reply frontend/src/lib/components/Table.tsx:10]: agreed, good shout!

File: frontend/src/lib/components/Table.tsx (hunk 4 above)
Comment (alexkim205) [reply frontend/src/lib/components/Table.tsx:23]: Got it let's do that! It doesn't look too complicated actually

File: frontend/src/lib/components/Table.tsx (hunk 3 above)
Comment (alexkim205) [reply frontend/src/lib/components/Table.tsx:10]: Something like this?

Screen Shot 2021-06-21 at 9 44 18 AM

Comment (mariusandra) [reply frontend/src/lib/components/Table.tsx:10]: Hmm, not really. I meant to just expand the column to fit everything... 🙃

Comment (alexkim205) [reply frontend/src/lib/components/Table.tsx:10]: Ah how does this look

Screen Shot 2021-06-21 at 10 40 53 AM

**Comment by alexkim205** [comment]
Addressed a-c above and added some goodies where the table doesn't word break columns when entering scroll mode. 

**Annotations Table**

https://user-images.githubusercontent.com/13460330/122806700-7de87d00-d27f-11eb-894f-6a933454af55.mov

**Feature Flags Table**

https://user-images.githubusercontent.com/13460330/122806731-8a6cd580-d27f-11eb-8860-6e972edea0b1.mov

**Word breaking for date-specific columns**
Screen Shot 2021-06-21 at 10 40 53 AM


P.S. Failing e2e tests don't seem related

**Comment by mariusandra** [comment]
Found (and fixed) just one weirdness with a double key for feature flags, probably from some merge:
image

Other than that, there's one new weirdness with the events table. Yet I'd suggest to ~~already~~finally merge this and tackle this in a new issue.

**Review by mariusandra — APPROVED** [review]

**Comment by mariusandra** [comment]
Even got it to look like this one time:
image

**Comment by alexkim205** [comment]
Thanks for the hot fix, looks like a bad merge resolution on my part. 

Yeah the events table (and other tables that use our custom ResizableTable component) is a bit funky and has a few bugs that I think should be addressed in another PR. `react-table` looks like the best approach so far for maintaining current table functionality while reducing antd gymnastics
